src/components/views/rooms/RoomHeader/CallGuestLinkButton.tsx

```diff
@@ -0,0 +1,189 @@
+                target,
+                customTitle: _t("share|share_call"),
+                subtitle: _t("share|share_call_subtitle"),
+            });
+        } catch (e) {
+            logger.error("Could not generate call link.", e);
+        }
+    }, [generateCallLink]);
+
+    const shareClick = useCallback(() => {
+        if (isRoomJoinable()) {
+            showLinkModal();
+        } else {
+            // the room needs to be set to public or knock to generate a link
+            Modal.createDialog(JoinRuleDialog, {
+                room,
+            }).finished.then(() => {
+                // we need to use the function here because the callback got called before the state was updated.
+                if (isRoomJoinable()) showLinkModal();
+            });
+        }
+    }, [showLinkModal, room, isRoomJoinable]);
+
+    return (
+        <>
+            {(canChangeJoinRule || isRoomJoinableState) && guestSpaUrl && (
+                <Tooltip label={_t("voip|get_call_link")}>
+                    <IconButton onClick={shareClick} aria-label={_t("voip|get_call_link")}>
+                        <ExternalLinkIcon />
+                    </IconButton>
+                </Tooltip>
+            )}
+        </>
+    );
+};
+
+/**
+ * A dialog to change the join rule of a room to public or knock.
+ * @param room The room to change the join rule of.
+ * @param onFinished Callback that is getting called if the dialog wants to close.
+ */
+export const JoinRuleDialog: React.FC<{
+    onFinished(): void;
+    room: Room;
+}> = ({ room, onFinished }) => {
+    const askToJoinEnabled = SettingsStore.getValue("feature_ask_to_join");
+    const [isUpdating, setIsUpdating] = React.useState<undefined | JoinRule>(undefined);
+    const changeJoinRule = useCallback(
+        async (newRule: JoinRule) => {
+            if (isUpdating !== undefined) return;
+            setIsUpdating(newRule);
+            await room.client.sendStateEvent(
+                room.roomId,
+                EventType.RoomJoinRules,
+                {
+                    join_rule: newRule,
+                } as IJoinRuleEventContent,
+                "",
+            );
+            // Show the dialog for a bit to give the user feedback
+            setTimeout(() => onFinished(), 500);
+        },
+        [isUpdating, onFinished, room.client, room.roomId],
+    );
+    return (
+        <BaseDialog title={_t("update_room_access_modal|title")} onFinished={onFinished}>
+            <div>
+                <p>{_t("update_room_access_modal|description")}</p>
+                {askToJoinEnabled && (
+                    <Button disabled={isUpdating === JoinRule.Knock} onClick={() => changeJoinRule(JoinRule.Knock)}>
+                        {_t("action|ask_to_join")}
+                    </Button>
+                )}
+                <Button
+                    // manually add destructive styles because they otherwise get overwritten by .mx_Dialog
```

Comment:
Actually this seems like a hack in general. From reading _common.pcss, it looks like you can give them the class `.mx_Dialog_nonDialogButton` to avoid the unwanted styling